scripts/TPI.py

```diff
@@ -0,0 +1,91 @@
+import numpy as np
+import scipy.optimize as opt
+
+def HHFOCs(b_sp1, *params):
+    '''
+    Refer to the equations on chapter 8
+    '''
+    BQpath, rho_s, beta, sigma, b_init, n, rpath, wpath = params
+    # rpath = length p, n = length p, p = number of periods left in life time, wpath = length p, b_sp1 = length p-1
+    
+    b_s = np.append(b_init, b_sp1)
+    b_sp1 = np.append(b_sp1, 0.0)
+    c = (1 + rpath) * b_s + wpath * n + BQpath - b_sp1
+    MU_c = c ** (-sigma)
+    errors = (MU_c[:-1] - beta * (1 + rpath[1:]) *
+              (1 - rho_s[:-1]) * MU_c[1:])
+    
+    return errors
+
+
+
+while dist > mindist and tpi_iter < maxiter:
+    
+    # define paths
+    b_11 = 1.1 * b_ss
+    BQ_params = (g_n_path[0], omega_S_preTP, rho_s)
+    K_params = (g_n_path[0], omega_S_preTP, imm_rates_path[0, :])
+    BQ_11 = get_BQ(b_11, r_11, BQ_params)
+    K_11 = get_K(b_11, K_params)
+    BQpath_init = np.zeros(T + S - 1)
+    Kpath_init = np.zeros(T + S - 1)
+    BQpath_init[:T] = np.linspace(BQ_11, BQ_ss, T)
+    Kpath_init[:T] = np.linspace(K_11, K_ss, T)
+    BQpath_init[T:] = BQ_ss
+    Kpath_init[T:] = K_ss
+    r_11 = get_r(K_11, L_ss, r_params)
+    rpath = get_r(Kpath_init, L_ss, r_params)
+    wpath = get_w(Kpath_init, L_ss, w_params)
+    bmat = np.zeros((S - 1, T + S - 1))
+    bmat[:, 0] = b_1
+    
+    tpi_iter += 1
+    dist = 8.0
+    mindist = 1e-08
+    maxiter = 300
+    tpi_iter = 0
+    xi = 0.2
+
+    # Solve for households
+    for p in range(2, S):
```

Comment:
This is the partial lifetime loop - @saharnazb, this relates to my comment on your PR.